# Preprocess PIC Database data

Beforee using ehrapy, we need to preprocess data and extract and merge the information we want and need.

First, we need to load the data into a pandas dataframe. To get an overview, we need informations on the icustay, the hospital admission and diagnoses.

In [1]:
import pandas as pd
import numpy as np
from IPython import display

In [2]:
# load the icd diagnoses definition table
icd_diagnoses_df = pd.read_csv("./V1.1.0/D_ICD_DIAGNOSES.csv")
# load the patient info
patients_df = pd.read_csv("./V1.1.0/PATIENTS.csv")
# load info on the hospital admissions
hadm_df = pd.read_csv("./V1.1.0/ADMISSIONS.csv")
# load info on the ICU stays
icu_stay_df = pd.read_csv("./V1.1.0/ICUSTAYS.csv")

# Explorative Data Analysis - Overview

In [3]:
patients_df.head()

ROW_ID  SUBJECT_ID GENDER                  DOB                  DOD  \
0       1           2      M  2068-11-30 00:00:00  2081-09-28 10:06:00   
1       2           9      M  2104-03-09 00:00:00                  NaN   
2       3          56      F  2110-01-08 13:40:00                  NaN   
3       4          79      F  2094-08-25 00:00:00                  NaN   
4       5          96      M  2050-12-17 00:00:00                  NaN   

   EXPIRE_FLAG  
0            1  
1            0  
2            0  
3            0  
4            0

In [4]:
hadm_df.head()

ROW_ID  SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME  \
0       1          26   100000  2098-11-09 18:30:55  2098-11-14 09:18:00   
1       2          28   100001  2104-09-03 10:36:47  2104-09-12 14:51:00   
2       3          29   100002  2062-11-29 20:52:52  2062-12-07 14:27:00   
3       4          30   100003  2098-11-21 15:38:47  2098-11-26 01:33:00   
4       5          31   100004  2071-11-13 19:43:13  2071-11-15 13:40:00   

             DEATHTIME ADMISSION_DEPARTMENT DISCHARGE_DEPARTMENT INSURANCE  \
0                  NaN          General ICU          General ICU  Self Pay   
1  2104-09-12 14:51:00          General ICU          General ICU  Self Pay   
2                  NaN          General ICU          General ICU  Self Pay   
3  2098-11-26 01:33:00          General ICU          General ICU  Discount   
4  2071-11-15 13:40:00          General ICU          General ICU  Self Pay   

  LANGUAGE       RELIGION MARITAL_STATUS   ETHNICITY            EDREGTIME  \
0      CHN  Not specified         Single  Han ethnic  2098-11-09 18:30:55   
1      CHN  Not specified         Single  Han ethnic  2104-09-03 10:36:47   
2      CHN  Not specified         Single  Han ethnic  2062-11-29 20:52:52   
3      CHN  Not specified         Single  Han ethnic  2098-11-21 15:38:47   
4      CHN  Not specified         Single  Han ethnic  2071-11-13 19:43:13   

             EDOUTTIME DIAGNOSIS ICD10_CODE_CN  HOSPITAL_EXPIRE_FLAG  \
0  2098-11-14 09:18:00      肠粘膜炎       K52.901                     0   
1  2104-09-12 14:51:00       心肌炎       I51.403                     1   
2  2062-12-07 14:27:00        溺水       T75.101                     0   
3  2098-11-26 01:33:00       NaN           NaN                     1   
4  2071-11-15 13:40:00       腹泻病       K52.904                     1   

   HAS_CHARTEVENTS_DATA  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0

In [5]:
icu_stay_df.head()

ROW_ID  SUBJECT_ID  HADM_ID  ICUSTAY_ID FIRST_CAREUNIT LAST_CAREUNIT  \
0       1          26   100000      200532    General ICU   General ICU   
1       2          28   100001      200533    General ICU   General ICU   
2       3          29   100002      200534    General ICU   General ICU   
3       4          30   100003      200535    General ICU   General ICU   
4       5          31   100004      200536    General ICU   General ICU   

   FIRST_WARDID  LAST_WARDID               INTIME              OUTTIME     LOS  
0           340          340  2098-11-09 18:30:55  2098-11-14 09:18:00  4.6167  
1           340          340  2104-09-03 10:36:46  2104-09-12 14:51:00  9.1771  
2           340          340  2062-11-29 20:52:52  2062-12-07 14:27:00  7.7326  
3           340          340  2098-11-21 15:38:47  2098-11-26 01:33:00  4.4132  
4           340          340  2071-11-13 19:43:12  2071-11-15 13:40:00  1.7479

In [6]:
icd_diagnoses_df.head()

ROW_ID ICD10_CODE_CN ICD10_CODE TITLE_CN  \
0       1       A00.000      A00.0  古典生物型霍乱   
1       2       A00.100      A00.1   埃尔托型霍乱   
2       3       A00.900      A00.9       霍乱   
3       4       A01.000      A01.0       伤寒   
4       5       A01.001      A01.0       伤寒   

                                               TITLE  
0  Cholera due to Vibrio cholerae 01, biovar chol...  
1    Cholera due to Vibrio cholerae 01, biovar eltor  
2                               Cholera, unspecified  
3                                      Typhoid fever  
4                                      Typhoid fever

In [7]:
# drop some columns, we won't ever use for now
hadm_df.drop(["ROW_ID", "DIAGNOSIS"], inplace=True, axis=1)
icu_stay_df.drop(["ROW_ID", "FIRST_WARDID", "LAST_WARDID"], inplace=True, axis=1)
icd_diagnoses_df.drop(["ROW_ID", "TITLE_CN"], inplace=True, axis=1)
patients_df.drop(["ROW_ID"], inplace=True, axis=1)

In [8]:
# check that they were dropped
hadm_df.head()

SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME  \
0          26   100000  2098-11-09 18:30:55  2098-11-14 09:18:00   
1          28   100001  2104-09-03 10:36:47  2104-09-12 14:51:00   
2          29   100002  2062-11-29 20:52:52  2062-12-07 14:27:00   
3          30   100003  2098-11-21 15:38:47  2098-11-26 01:33:00   
4          31   100004  2071-11-13 19:43:13  2071-11-15 13:40:00   

             DEATHTIME ADMISSION_DEPARTMENT DISCHARGE_DEPARTMENT INSURANCE  \
0                  NaN          General ICU          General ICU  Self Pay   
1  2104-09-12 14:51:00          General ICU          General ICU  Self Pay   
2                  NaN          General ICU          General ICU  Self Pay   
3  2098-11-26 01:33:00          General ICU          General ICU  Discount   
4  2071-11-15 13:40:00          General ICU          General ICU  Self Pay   

  LANGUAGE       RELIGION MARITAL_STATUS   ETHNICITY            EDREGTIME  \
0      CHN  Not specified         Single  Han ethnic  2098-11-09 18:30:55   
1      CHN  Not specified         Single  Han ethnic  2104-09-03 10:36:47   
2      CHN  Not specified         Single  Han ethnic  2062-11-29 20:52:52   
3      CHN  Not specified         Single  Han ethnic  2098-11-21 15:38:47   
4      CHN  Not specified         Single  Han ethnic  2071-11-13 19:43:13   

             EDOUTTIME ICD10_CODE_CN  HOSPITAL_EXPIRE_FLAG  \
0  2098-11-14 09:18:00       K52.901                     0   
1  2104-09-12 14:51:00       I51.403                     1   
2  2062-12-07 14:27:00       T75.101                     0   
3  2098-11-26 01:33:00           NaN                     1   
4  2071-11-15 13:40:00       K52.904                     1   

   HAS_CHARTEVENTS_DATA  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0

In [9]:
# general overview of PATIENTS data
display.display(patients_df.info(show_counts=1))
display.display(patients_df.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12881 entries, 0 to 12880
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   SUBJECT_ID   12881 non-null  int64 
 1   GENDER       12881 non-null  object
 2   DOB          12881 non-null  object
 3   DOD          960 non-null    object
 4   EXPIRE_FLAG  12881 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 503.3+ KB


None

SUBJECT_ID   EXPIRE_FLAG
count  12881.000000  12881.000000
mean    6704.959009      0.074528
std     3910.277034      0.262639
min        2.000000      0.000000
25%     3358.000000      0.000000
50%     6682.000000      0.000000
75%    10079.000000      0.000000
max    13643.000000      1.000000

No missing data! About 7.5% death rate.

In [10]:
# general overview of ADMISSION data
display.display(hadm_df.info(show_counts=1))
display.display(hadm_df.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13449 entries, 0 to 13448
Data columns (total 17 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   SUBJECT_ID            13449 non-null  int64 
 1   HADM_ID               13449 non-null  int64 
 2   ADMITTIME             13449 non-null  object
 3   DISCHTIME             13449 non-null  object
 4   DEATHTIME             971 non-null    object
 5   ADMISSION_DEPARTMENT  13449 non-null  object
 6   DISCHARGE_DEPARTMENT  13449 non-null  object
 7   INSURANCE             13449 non-null  object
 8   LANGUAGE              13449 non-null  object
 9   RELIGION              13449 non-null  object
 10  MARITAL_STATUS        13449 non-null  object
 11  ETHNICITY             13449 non-null  object
 12  EDREGTIME             13449 non-null  object
 13  EDOUTTIME             13449 non-null  object
 14  ICD10_CODE_CN         13372 non-null  object
 15  HOSPITAL_EXPIRE_FLAG  13449 non-null

None

SUBJECT_ID        HADM_ID  HOSPITAL_EXPIRE_FLAG  HAS_CHARTEVENTS_DATA
count  13449.000000   13449.000000          13449.000000          13449.000000
mean    6688.591643  106996.615659              0.072199              0.797085
std     3883.761314    4080.919919              0.258826              0.402184
min        2.000000  100000.000000              0.000000              0.000000
25%     3394.000000  103502.000000              0.000000              1.000000
50%     6650.000000  106976.000000              0.000000              1.000000
75%    10022.000000  110529.000000              0.000000              1.000000
max    13643.000000  114202.000000              1.000000              1.000000

Small oddity here: 971 deathtimes vs 960 total deaths in PATIENTS? 

About 79,7% of admissions have acutally event data!

About 99,4% have a diagnosis!

In [11]:
# general overview of ICUSTAYS data
display.display(icu_stay_df.info(show_counts=1))
display.display(icu_stay_df.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13941 entries, 0 to 13940
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   SUBJECT_ID      13941 non-null  int64  
 1   HADM_ID         13941 non-null  int64  
 2   ICUSTAY_ID      13941 non-null  int64  
 3   FIRST_CAREUNIT  13941 non-null  object 
 4   LAST_CAREUNIT   13941 non-null  object 
 5   INTIME          13941 non-null  object 
 6   OUTTIME         13941 non-null  object 
 7   LOS             13941 non-null  float64
dtypes: float64(1), int64(3), object(4)
memory usage: 871.4+ KB


None

SUBJECT_ID        HADM_ID     ICUSTAY_ID           LOS
count  13941.000000   13941.000000   13941.000000  13941.000000
mean    6724.538986  107036.252564  207511.508428      9.302988
std     3865.567024    4061.140715    4028.106482     18.635520
min        2.000000  100000.000000  200532.000000     -0.470100
25%     3447.000000  103562.000000  204027.000000      0.921500
50%     6713.000000  107051.000000  207514.000000      2.802100
75%    10040.000000  110542.000000  210999.000000      9.169400
max    13643.000000  114202.000000  214513.000000    444.045100

No missing data!

In [12]:
# general overview of D_DIAGNOSIS data
display.display(icd_diagnoses_df.info(show_counts=1))
display.display(icd_diagnoses_df.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25379 entries, 0 to 25378
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   ICD10_CODE_CN  25379 non-null  object
 1   ICD10_CODE     25377 non-null  object
 2   TITLE          25379 non-null  object
dtypes: object(3)
memory usage: 594.9+ KB


None

ICD10_CODE_CN ICD10_CODE                             TITLE
count          25379      25377                             25379
unique         25379      10038                              9779
top          A00.000      A18.0  Tuberculosis of bones and joints
freq               1         76                                76

2 ICD codes are missing (non CN) might be no big deal.

# Map diagnosis to CN_ICD per HADM
To start, we want to get a feeling for the diseases were dealing with in the data. Therefore, get their names from the ICD_DIAGNOSES definition table.

Ignoring DIAGNOSES_ICD for now, we're only using the admission diagnosis here, but we will address this issue later.

In [13]:
# key value pairs of cn_icd code and disease name
icd_to_disease = {
    icd_cn: disease
    for icd_cn, disease in zip(
        icd_diagnoses_df["ICD10_CODE_CN"], icd_diagnoses_df["TITLE"]
    )
}

In [14]:
hadm_df["disease_name"] = hadm_df["ICD10_CODE_CN"].map(icd_to_disease)

In [15]:
hadm_df.head()

SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME  \
0          26   100000  2098-11-09 18:30:55  2098-11-14 09:18:00   
1          28   100001  2104-09-03 10:36:47  2104-09-12 14:51:00   
2          29   100002  2062-11-29 20:52:52  2062-12-07 14:27:00   
3          30   100003  2098-11-21 15:38:47  2098-11-26 01:33:00   
4          31   100004  2071-11-13 19:43:13  2071-11-15 13:40:00   

             DEATHTIME ADMISSION_DEPARTMENT DISCHARGE_DEPARTMENT INSURANCE  \
0                  NaN          General ICU          General ICU  Self Pay   
1  2104-09-12 14:51:00          General ICU          General ICU  Self Pay   
2                  NaN          General ICU          General ICU  Self Pay   
3  2098-11-26 01:33:00          General ICU          General ICU  Discount   
4  2071-11-15 13:40:00          General ICU          General ICU  Self Pay   

  LANGUAGE       RELIGION MARITAL_STATUS   ETHNICITY            EDREGTIME  \
0      CHN  Not specified         Single  Han ethnic  2098-11-09 18:30:55   
1      CHN  Not specified         Single  Han ethnic  2104-09-03 10:36:47   
2      CHN  Not specified         Single  Han ethnic  2062-11-29 20:52:52   
3      CHN  Not specified         Single  Han ethnic  2098-11-21 15:38:47   
4      CHN  Not specified         Single  Han ethnic  2071-11-13 19:43:13   

             EDOUTTIME ICD10_CODE_CN  HOSPITAL_EXPIRE_FLAG  \
0  2098-11-14 09:18:00       K52.901                     0   
1  2104-09-12 14:51:00       I51.403                     1   
2  2062-12-07 14:27:00       T75.101                     0   
3  2098-11-26 01:33:00           NaN                     1   
4  2071-11-15 13:40:00       K52.904                     1   

   HAS_CHARTEVENTS_DATA                                       disease_name  
0                     0  Noninfective gastroenteritis and colitis, unsp...  
1                     0                           Myocarditis, unspecified  
2                     0                   Drowning and nonfatal submersion  
3                     0                                                NaN  
4                     0  Noninfective gastroenteritis and colitis, unsp...

Now, we need to do the same thing to map a disease from hadm_df to icustays!

In [16]:
hadm_id_to_disease = {
    hadm_id: disease
    for hadm_id, disease in zip(hadm_df["HADM_ID"], hadm_df["disease_name"])
}

In [17]:
icu_stay_df["disease_name"] = icu_stay_df["HADM_ID"].map(hadm_id_to_disease)

In [18]:
icu_stay_df.head()

SUBJECT_ID  HADM_ID  ICUSTAY_ID FIRST_CAREUNIT LAST_CAREUNIT  \
0          26   100000      200532    General ICU   General ICU   
1          28   100001      200533    General ICU   General ICU   
2          29   100002      200534    General ICU   General ICU   
3          30   100003      200535    General ICU   General ICU   
4          31   100004      200536    General ICU   General ICU   

                INTIME              OUTTIME     LOS  \
0  2098-11-09 18:30:55  2098-11-14 09:18:00  4.6167   
1  2104-09-03 10:36:46  2104-09-12 14:51:00  9.1771   
2  2062-11-29 20:52:52  2062-12-07 14:27:00  7.7326   
3  2098-11-21 15:38:47  2098-11-26 01:33:00  4.4132   
4  2071-11-13 19:43:12  2071-11-15 13:40:00  1.7479   

                                        disease_name  
0  Noninfective gastroenteritis and colitis, unsp...  
1                           Myocarditis, unspecified  
2                   Drowning and nonfatal submersion  
3                                                NaN  
4  Noninfective gastroenteritis and colitis, unsp...

In [19]:
# round icu stay time (days) since we do not want to deal with "floating" days
icu_stay_df["LOS"] = icu_stay_df["LOS"].round()

In [20]:
icu_stay_df.head()

SUBJECT_ID  HADM_ID  ICUSTAY_ID FIRST_CAREUNIT LAST_CAREUNIT  \
0          26   100000      200532    General ICU   General ICU   
1          28   100001      200533    General ICU   General ICU   
2          29   100002      200534    General ICU   General ICU   
3          30   100003      200535    General ICU   General ICU   
4          31   100004      200536    General ICU   General ICU   

                INTIME              OUTTIME  LOS  \
0  2098-11-09 18:30:55  2098-11-14 09:18:00  5.0   
1  2104-09-03 10:36:46  2104-09-12 14:51:00  9.0   
2  2062-11-29 20:52:52  2062-12-07 14:27:00  8.0   
3  2098-11-21 15:38:47  2098-11-26 01:33:00  4.0   
4  2071-11-13 19:43:12  2071-11-15 13:40:00  2.0   

                                        disease_name  
0  Noninfective gastroenteritis and colitis, unsp...  
1                           Myocarditis, unspecified  
2                   Drowning and nonfatal submersion  
3                                                NaN  
4  Noninfective gastroenteritis and colitis, unsp...

# Add disease hierarchy
First of all, we want to know which type of diseases were dealing with (for example respiratory diseases or infectious diseases). Since ehrapy does not currently offer an ICD mapper (WIP!), do this manually for now. 


In [21]:
icd_diagnoses_df["disease_type"] = "Unknown"
icd_diagnoses_df.loc[
    ("A00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "B99"),
    "disease_type",
] = "Infection"
icd_diagnoses_df.loc[
    ("C00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "D48"),
    "disease_type",
] = "Neoplasm"
icd_diagnoses_df.loc[
    ("D50" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "D89"),
    "disease_type",
] = "Blood and Immune system"
icd_diagnoses_df.loc[
    ("E00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "E90"),
    "disease_type",
] = "Metabolism and Endocrine"
icd_diagnoses_df.loc[
    ("F00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "F99"),
    "disease_type",
] = "Psych"
icd_diagnoses_df.loc[
    ("G00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "G99"),
    "disease_type",
] = "Nervous system"
icd_diagnoses_df.loc[
    ("H00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "H59"),
    "disease_type",
] = "Eye"
icd_diagnoses_df.loc[
    ("H60" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "H95"),
    "disease_type",
] = "Ear"
icd_diagnoses_df.loc[
    ("I00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "I99"),
    "disease_type",
] = "Cardiovascular"
icd_diagnoses_df.loc[
    ("J00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "J99"),
    "disease_type",
] = "Respiratory"
icd_diagnoses_df.loc[
    ("K00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "K93"),
    "disease_type",
] = "Digestive system"
icd_diagnoses_df.loc[
    ("L00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "L99"),
    "disease_type",
] = "Skin"
icd_diagnoses_df.loc[
    ("M00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "M99"),
    "disease_type",
] = "Musculosceletal"
icd_diagnoses_df.loc[
    ("N00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "N99"),
    "disease_type",
] = "Urogenital system"
icd_diagnoses_df.loc[
    ("O00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "O99"),
    "disease_type",
] = "Prae-/Neonatal"
icd_diagnoses_df.loc[
    ("P00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "P96"),
    "disease_type",
] = "Perinatal"
icd_diagnoses_df.loc[
    ("Q00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "Q99"),
    "disease_type",
] = "Malformations"
icd_diagnoses_df.loc[
    ("R00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "R99"),
    "disease_type",
] = "Other"
icd_diagnoses_df.loc[
    ("S00" <= icd_diagnoses_df["ICD10_CODE"])
    & (icd_diagnoses_df["ICD10_CODE"] <= "Y98"),
    "disease_type",
] = "Posoning, Injury or External"

In [22]:
icd_diagnoses_df.head()

ICD10_CODE_CN ICD10_CODE                                              TITLE  \
0       A00.000      A00.0  Cholera due to Vibrio cholerae 01, biovar chol...   
1       A00.100      A00.1    Cholera due to Vibrio cholerae 01, biovar eltor   
2       A00.900      A00.9                               Cholera, unspecified   
3       A01.000      A01.0                                      Typhoid fever   
4       A01.001      A01.0                                      Typhoid fever   

  disease_type  
0    Infection  
1    Infection  
2    Infection  
3    Infection  
4    Infection

In [23]:
# map disease type to HADM TODO:
icd_to_disease_type = {
    icd_cn: disease_type
    for icd_cn, disease_type in zip(
        icd_diagnoses_df["ICD10_CODE_CN"], icd_diagnoses_df["disease_type"]
    )
}
hadm_df["disease_type"] = hadm_df["ICD10_CODE_CN"].map(icd_to_disease_type)

In [24]:
hadm_df.head()

SUBJECT_ID  HADM_ID            ADMITTIME            DISCHTIME  \
0          26   100000  2098-11-09 18:30:55  2098-11-14 09:18:00   
1          28   100001  2104-09-03 10:36:47  2104-09-12 14:51:00   
2          29   100002  2062-11-29 20:52:52  2062-12-07 14:27:00   
3          30   100003  2098-11-21 15:38:47  2098-11-26 01:33:00   
4          31   100004  2071-11-13 19:43:13  2071-11-15 13:40:00   

             DEATHTIME ADMISSION_DEPARTMENT DISCHARGE_DEPARTMENT INSURANCE  \
0                  NaN          General ICU          General ICU  Self Pay   
1  2104-09-12 14:51:00          General ICU          General ICU  Self Pay   
2                  NaN          General ICU          General ICU  Self Pay   
3  2098-11-26 01:33:00          General ICU          General ICU  Discount   
4  2071-11-15 13:40:00          General ICU          General ICU  Self Pay   

  LANGUAGE       RELIGION MARITAL_STATUS   ETHNICITY            EDREGTIME  \
0      CHN  Not specified         Single  Han ethnic  2098-11-09 18:30:55   
1      CHN  Not specified         Single  Han ethnic  2104-09-03 10:36:47   
2      CHN  Not specified         Single  Han ethnic  2062-11-29 20:52:52   
3      CHN  Not specified         Single  Han ethnic  2098-11-21 15:38:47   
4      CHN  Not specified         Single  Han ethnic  2071-11-13 19:43:13   

             EDOUTTIME ICD10_CODE_CN  HOSPITAL_EXPIRE_FLAG  \
0  2098-11-14 09:18:00       K52.901                     0   
1  2104-09-12 14:51:00       I51.403                     1   
2  2062-12-07 14:27:00       T75.101                     0   
3  2098-11-26 01:33:00           NaN                     1   
4  2071-11-15 13:40:00       K52.904                     1   

   HAS_CHARTEVENTS_DATA                                       disease_name  \
0                     0  Noninfective gastroenteritis and colitis, unsp...   
1                     0                           Myocarditis, unspecified   
2                     0                   Drowning and nonfatal submersion   
3                     0                                                NaN   
4                     0  Noninfective gastroenteritis and colitis, unsp...   

                   disease_type  
0              Digestive system  
1                Cardiovascular  
2  Posoning, Injury or External  
3                           NaN  
4              Digestive system

In [25]:
# same again for HADM to ICUSTAY
hadm_id_to_disease_type = {
    hadm_id: disease_type
    for hadm_id, disease_type in zip(hadm_df["HADM_ID"], hadm_df["disease_type"])
}
icu_stay_df["disease_type"] = icu_stay_df["HADM_ID"].map(hadm_id_to_disease_type)

In [26]:
icu_stay_df.head()

SUBJECT_ID  HADM_ID  ICUSTAY_ID FIRST_CAREUNIT LAST_CAREUNIT  \
0          26   100000      200532    General ICU   General ICU   
1          28   100001      200533    General ICU   General ICU   
2          29   100002      200534    General ICU   General ICU   
3          30   100003      200535    General ICU   General ICU   
4          31   100004      200536    General ICU   General ICU   

                INTIME              OUTTIME  LOS  \
0  2098-11-09 18:30:55  2098-11-14 09:18:00  5.0   
1  2104-09-03 10:36:46  2104-09-12 14:51:00  9.0   
2  2062-11-29 20:52:52  2062-12-07 14:27:00  8.0   
3  2098-11-21 15:38:47  2098-11-26 01:33:00  4.0   
4  2071-11-13 19:43:12  2071-11-15 13:40:00  2.0   

                                        disease_name  \
0  Noninfective gastroenteritis and colitis, unsp...   
1                           Myocarditis, unspecified   
2                   Drowning and nonfatal submersion   
3                                                NaN   
4  Noninfective gastroenteritis and colitis, unsp...   

                   disease_type  
0              Digestive system  
1                Cardiovascular  
2  Posoning, Injury or External  
3                           NaN  
4              Digestive system

In [27]:
icu_stay_df["disease_type"].unique()

array(['Digestive system', 'Cardiovascular',
       'Posoning, Injury or External', nan, 'Respiratory',
       'Malformations', 'Infection', 'Nervous system',
       'Metabolism and Endocrine', 'Other', 'Psych', 'Neoplasm',
       'Perinatal', 'Blood and Immune system', 'Unknown',
       'Urogenital system', 'Musculosceletal', 'Skin', 'Prae-/Neonatal',
       'Eye', 'Ear'], dtype=object)

In [28]:
# add gender for each patient
patient_id_to_gender = {
    patient_id: gender
    for patient_id, gender in zip(patients_df["SUBJECT_ID"], patients_df["GENDER"])
}
icu_stay_df["gender"] = icu_stay_df["SUBJECT_ID"].map(patient_id_to_gender)

In [29]:
# add some other demographic features for each patient
for feature in ["INSURANCE", "LANGUAGE", "RELIGION", "MARITAL_STATUS", "ETHNICITY"]:
    hadm_id_to_feature = {
        hadm_id: feature
        for hadm_id, feature in zip(hadm_df["HADM_ID"], hadm_df[feature])
    }
    icu_stay_df[feature.lower()] = icu_stay_df["HADM_ID"].map(hadm_id_to_feature)

In [30]:
# death during hospitalization?
hadm_id_to_death_flg = {
    hadm_id: death_flg
    for hadm_id, death_flg in zip(hadm_df["HADM_ID"], hadm_df["HOSPITAL_EXPIRE_FLAG"])
}
icu_stay_df["Death"] = icu_stay_df["HADM_ID"].map(hadm_id_to_death_flg)

## Labevents
Add most common blood values as well. To reduce influence of missing data and the different blood values measured different times, we will use the following strategy:

* Get the most common blood values across all HADM IDs (threshold should be ideally above 80-90% all values included in the inital anaylsis are measured at least once during the specific hospital stay). Filter all values, that are clinically relevant across a large population of ICU patients. This includes renal, liver and BGA parameters.

* Include the max, min and average measurement for each such value + a flag, that indicates whether a value is in the normal range of values or not.

* WIP: Include a trend parameter.

In [31]:
lab_events_df = pd.read_csv("./V1.1.0/LABEVENTS.csv")
ref_lab_values = pd.read_csv("./REFERENCE_VALUES_LAB.csv")

### First step: Find the **n most common values** measured across the HADMs.

In [32]:
lab_grp_by_hadm = lab_events_df.groupby("ITEMID")["HADM_ID"].nunique()

In [33]:
ref_lab_values["TEST_DETAIL_ITEM_CODE"] = pd.to_numeric(
    ref_lab_values["TEST_DETAIL_ITEM_CODE"], errors="coerce", downcast="integer"
)
ref_lab_values = ref_lab_values[~ref_lab_values["TEST_DETAIL_ITEM_CODE"].isna()]
ref_lab_values["TEST_DETAIL_ITEM_CODE"] = ref_lab_values[
    "TEST_DETAIL_ITEM_CODE"
].astype("int32")

In [34]:
ref_lab_values.dtypes

TEST_DETAIL_ITEM_CODE     int32
TEST_DETAIL_ITEM_NAME    object
REFERENCE_VALUE          object
dtype: object

In [35]:
# prepare ref lab value dataframe for easier access afterwards
ref_lab_values.drop("TEST_DETAIL_ITEM_NAME", inplace=True, axis=1)
ref_lab_values.set_index("TEST_DETAIL_ITEM_CODE", inplace=True)
ref_lab_values.head(2)

REFERENCE_VALUE
TEST_DETAIL_ITEM_CODE                
5002                          1.0-6.0
5004                         Negative

In [36]:
# side note: there are 13449 different HADMs in the dataset
# so check how many values had at least one measurement relative to all HADMs
most_common_values = lab_grp_by_hadm.loc[(lab_grp_by_hadm / 13449) > 0.90]

In [37]:
# for a quick clinical check, check how those values are called
d_labitem_df = pd.read_csv("./V1.1.0/D_LABITEMS.csv")
idx_lst = most_common_values.index.to_list()
id_to_name_grp = d_labitem_df.query("ITEMID in @idx_lst").groupby("ITEMID")["LABEL"]

In [38]:
# map name to item_id; this is possible as one item_id always defines one single value and not multiples
id_to_name = id_to_name_grp.apply(
    lambda grp: list(grp.value_counts().index)[0]
).to_dict()

In [39]:
list(id_to_name.keys())

[5002,
 5021,
 5022,
 5024,
 5025,
 5026,
 5027,
 5031,
 5033,
 5034,
 5037,
 5038,
 5039,
 5041,
 5042,
 5045,
 5046,
 5055,
 5057,
 5072,
 5074,
 5075,
 5077,
 5078,
 5083,
 5094,
 5095,
 5097,
 5099,
 5110,
 5111,
 5113,
 5114,
 5115,
 5120,
 5127,
 5128,
 5129,
 5132,
 5136,
 5141,
 5211,
 5215,
 5218,
 5219,
 5223,
 5224,
 5225,
 5226,
 5227,
 5228,
 5230,
 5235,
 5237,
 5239,
 5248,
 5249,
 5252,
 5257,
 5306,
 5492,
 5493,
 6236,
 6261,
 6316,
 6317,
 6318,
 6471,
 6472,
 6473]

In [40]:
ref_lab_values = ref_lab_values[ref_lab_values.index.isin(id_to_name.keys())]

In [41]:
ref_lab_values = ref_lab_values[~ref_lab_values["REFERENCE_VALUE"].isna()]

In [42]:
ref_lab_values

REFERENCE_VALUE
TEST_DETAIL_ITEM_CODE                
5002                          1.0-6.0
5024                        32.0-52.0
5041                          2.55-20
5041                       44.0-120.0
5211                               -6
...                               ...
6316                        0.10-1.50
6261                            <1.10
6317                         4.0-16.0
6261                      0.630-1.200
6318                        0.05-0.50

[199 rows x 1 columns]

In [43]:
print(
    f"We have {len(list(id_to_name.keys()))} lab values and {len(list(ref_lab_values.index.unique()))} ref values"
)

We have 70 lab values and 68 ref values


In [44]:
# which ones are missing?
diff = set(id_to_name.keys()) - set(ref_lab_values.index.unique())
print(f"Missing: {', '.join(id_to_name[item_id] for item_id in diff)}")

Missing: Serum  icteric index, Lipase


In [45]:
for item_id in list(id_to_name.keys()):
    lab_name = id_to_name[item_id]
    kwargs_dict = {
        f"{lab_name}_min": ("VALUENUM", np.min),
        f"{lab_name}_max": ("VALUENUM", np.max),
        f"{lab_name}_avg": ("VALUENUM", np.nanmean),
    }
    lab_events_df_temp = lab_events_df[lab_events_df["ITEMID"] == item_id]
    max_lab_value_per_hadm = lab_events_df_temp.groupby("HADM_ID").agg(**kwargs_dict)
    # todo check why this reduces size of dataframe (removes rows and columns)
    # icu_stay_df = pd.merge(icu_stay_df, max_lab_value_per_hadm, on="HADM_ID")
    for col in kwargs_dict.keys():
        hadm_id_to_col = {
            hadm_id: col
            for hadm_id, col in zip(
                max_lab_value_per_hadm.index, max_lab_value_per_hadm[col]
            )
        }
        icu_stay_df[col] = icu_stay_df["HADM_ID"].map(hadm_id_to_col)
        # round avg column: 10,12345 -> 10,12
        if col.endswith("_avg"):
            icu_stay_df[col] = icu_stay_df[col].round(2)
    # flag to indicate whether a lab value has been measured or not
    # this works, because whenever a value has no minimum (or maximum or avg) in this row, it has not been measured at all
    # icu_stay_df[f'{lab_name}_measured'] = np.where(icu_stay_df[f'{lab_name}_min'] == , 0, 1)

/tmp/ipykernel_32177/2413784329.py:10: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  icu_stay_df[col] = icu_stay_df['HADM_ID'].map(hadm_id_to_col)
/tmp/ipykernel_32177/2413784329.py:10: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  icu_stay_df[col] = icu_stay_df['HADM_ID'].map(hadm_id_to_col)
/tmp/ipykernel_32177/2413784329.py:10: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(a

In [47]:
icu_stay_df_merged = pd.merge(icu_stay_df, patients_df, on="SUBJECT_ID").drop(
    ["GENDER", "EXPIRE_FLAG", "DOD"], axis=1
)

In [48]:
icu_stay_df_merged

SUBJECT_ID  HADM_ID  ICUSTAY_ID FIRST_CAREUNIT LAST_CAREUNIT  \
0              26   100000      200532    General ICU   General ICU   
1              28   100001      200533    General ICU   General ICU   
2              29   100002      200534    General ICU   General ICU   
3              30   100003      200535    General ICU   General ICU   
4              31   100004      200536    General ICU   General ICU   
...           ...      ...         ...            ...           ...   
13936       13423   114192      214506           CICU          CICU   
13937       13639   114198      214510           SICU          SICU   
13938       13642   114199      214511           SICU          SICU   
13939       13643   114200      214512           SICU          SICU   
13940       12553   114202      214513           SICU          SICU   

                    INTIME              OUTTIME  LOS  \
0      2098-11-09 18:30:55  2098-11-14 09:18:00  5.0   
1      2104-09-03 10:36:46  2104-09-12 14:51:00  9.0   
2      2062-11-29 20:52:52  2062-12-07 14:27:00  8.0   
3      2098-11-21 15:38:47  2098-11-26 01:33:00  4.0   
4      2071-11-13 19:43:12  2071-11-15 13:40:00  2.0   
...                    ...                  ...  ...   
13936  2110-02-04 12:06:58  2110-02-08 09:25:03  4.0   
13937  2072-02-07 17:08:10  2072-02-09 11:30:00  2.0   
13938  2100-02-23 15:16:46  2100-02-24 12:29:01  1.0   
13939  2090-02-21 13:51:05  2090-02-21 18:14:53  0.0   
13940  2115-12-10 10:43:29  2115-12-10 16:30:21  0.0   

                                            disease_name  \
0      Noninfective gastroenteritis and colitis, unsp...   
1                               Myocarditis, unspecified   
2                       Drowning and nonfatal submersion   
3                                                    NaN   
4      Noninfective gastroenteritis and colitis, unsp...   
...                                                  ...   
13936                          Ventricular septal defect   
13937                               Compression of brain   
13938                                   Craniosynostosis   
13939  Other specified congenital malformations of sp...   
13940  Other specified congenital malformations of sp...   

                       disease_type  ... Serum hemolytic index_min  \
0                  Digestive system  ...                       NaN   
1                    Cardiovascular  ...                       NaN   
2      Posoning, Injury or External  ...                       NaN   
3                               NaN  ...                       NaN   
4                  Digestive system  ...                       NaN   
...                             ...  ...                       ...   
13936                 Malformations  ...                       0.0   
13937                Nervous system  ...                       0.0   
13938                 Malformations  ...                       0.0   
13939                 Malformations  ...                       0.0   
13940                 Malformations  ...                       0.0   

      Serum hemolytic index_max Serum hemolytic index_avg  \
0                           NaN                       NaN   
1                           NaN                       NaN   
2                           NaN                       NaN   
3                           NaN                       NaN   
4                           NaN                       NaN   
...                         ...                       ...   
13936                       0.0                       0.0   
13937                       0.0                       0.0   
13938                       0.0                       0.0   
13939                       0.0                       0.0   
13940                       0.0                       0.0   

      Serum  icteric index_min Serum  icteric index_max  \
0                          NaN                      NaN   
1                          NaN                      NaN   
2        

In [49]:
icu_stay_df_merged["DOB"] = pd.to_datetime(icu_stay_df_merged["DOB"])
icu_stay_df_merged["INTIME"] = pd.to_datetime(icu_stay_df_merged["INTIME"])

In [50]:
# calculate age on a per month basis; therefore we need to divide by 30 (sufficient estimate)
def calc_age(now, dob):
    return ((now - dob) / 30).astype(int)

In [51]:
icu_stay_df_merged["age_in_month"] = calc_age(
    icu_stay_df_merged["INTIME"].to_numpy().astype("datetime64[D]"),
    icu_stay_df_merged["DOB"].to_numpy().astype("datetime64[D]"),
)

In [52]:
icu_stay_df_merged.drop("DOB", inplace=True, axis=1)

In [55]:
icu_stay_df_merged.to_csv("./merged_icustay_no_symptoms.csv", index=False)

Now let's add symptoms to the resulting dataframe. To reduce dimensionality, we are starting by grouping the different symptoms into categories grouped by their disease complex. So if a patient showed a specific smyptom at any time during its ICU stay, this will be marked as present, otherwise as absent (could be enhanced by subgrouping into always present, mostly present, mostly absent always absent, for example).

__Example__:
`heart murmur` and `cyanosis` are both symptoms especially from the cardiovascular disease complex.
Note: Some symptoms may fall into multiple categories.
Symptoms like `nausea` or `pain` (if not specified further) will be grouped into a `vegetative/general` symptom group.

In [55]:
symptoms_df = pd.read_csv("./V1.1.0/EMR_SYMPTOMS.csv")

In [56]:
symptoms_df.drop(
    ["ROW_ID", "EMR_ID", "RECORDTIME", "SYMPTOM_NAME_CN"], axis=1, inplace=True
)

In [57]:
symptoms_df.head()

SUBJECT_ID  HADM_ID          SYMPTOM_NAME SYMPTOM_ATTRIBUTE
0       12104   112523                 cough                 +
1       12104   112523         good appetite                 +
2       12104   112523           good spirit                 +
3       12104   112523              listless                 +
4       12104   112523  regular heart rhythm                 +

In [58]:
# filter only present symptoms for now
symptoms_df = symptoms_df[symptoms_df["SYMPTOM_ATTRIBUTE"] == "+"]

In [59]:
# check all unique symptoms
pd.unique(symptoms_df["SYMPTOM_NAME"])  # total of 335 different symptoms

array(['cough', 'good appetite', 'good spirit', 'listless',
       'regular heart rhythm', 'abdominal soft', 'ecchymosis',
       'cool extremities', 'heart murmur',
       'ventricular septal defect (VSD)', 'obnubilation',
       'sleep peacefully', 'pupils equal and round',
       'congestion\xa0of\xa0throat', 'smooth breathing', 'fever',
       'soft neck',
       'swelling\xa0of the\xa0spleen\xa0under\xa0the\xa0ribs',
       'phlegm sound', 'pharyngeal red', 'swelling',
       'bilateral\xa0swelling\xa0of the\xa0neck',
       'idiopathic sinus tachycardia (IST)', 'twitching',
       'Patent ductus arteriosus(PDA)', 'patent foramen ovale (PFO)',
       'atelectasis\xa0in\xa0the\xa0right upper lung',
       'liver dysfunction', 'blue jaundice of lips',
       'cardiopulmonary\xa0arrest', 'vascular nerve injury', 'hemorrhage',
       'cardiac\xa0arrest', 'surgical\xa0incision\xa0necrosis',
       'tissue necrosis', 'severe\xa0systemic\xa0disease', 'defect',
       'headache', 'infecti

In [60]:
# this is needed in order to remove `\xa0` from the values in SYMPTOM_NAMES
# `\xa0` is actually non-breaking space in Latin1 (ISO 8859-1),
import unicodedata


def preprocess_for_normalize(x):
    # Convert back to ascii
    return unicodedata.normalize("NFKD", x).encode("ascii", "ignore").decode("utf-8")

In [61]:
symptoms_df["SYMPTOM_NAME"] = symptoms_df["SYMPTOM_NAME"].apply(
    lambda x: preprocess_for_normalize(x) if (pd.notnull(x)) else x
)

In [62]:
pd.unique(symptoms_df["SYMPTOM_NAME"])

array(['cough', 'good appetite', 'good spirit', 'listless',
       'regular heart rhythm', 'abdominal soft', 'ecchymosis',
       'cool extremities', 'heart murmur',
       'ventricular septal defect (VSD)', 'obnubilation',
       'sleep peacefully', 'pupils equal and round',
       'congestion of throat', 'smooth breathing', 'fever', 'soft neck',
       'swelling of the spleen under the ribs', 'phlegm sound',
       'pharyngeal red', 'swelling', 'bilateral swelling of the neck',
       'idiopathic sinus tachycardia (IST)', 'twitching',
       'Patent ductus arteriosus(PDA)', 'patent foramen ovale (PFO)',
       'atelectasis in the right upper lung', 'liver dysfunction',
       'blue jaundice of lips', 'cardiopulmonary arrest',
       'vascular nerve injury', 'hemorrhage', 'cardiac arrest',
       'surgical incision necrosis', 'tissue necrosis',
       'severe systemic disease', 'defect', 'headache', 'infection',
       'necrosis of skin and extremity', 'severe infection',
       'low-

In [63]:
gastrointestinal_symptoms = [
    "good appetite",
    "abdominal soft",
    "general appetite",
    "liver dysfunction",
    "abdominal pain",
    "abdomen flat",
    "borborygmus",
    "abdominal bulge",
    "diarrhea",
    "ascites",
    "hepatic coma",
    "jaundice",
    "ano-rectal malformation",
    "skin jaundice",
    "ascites",
    "changes in the abdomen",
    "abdominal distension",
    "hematochezia",
    "inflatable stomach intestine",
    "milk regurgitation",
    "abnormality of abdomen",
    "changes in the abdomen",
    "vomiting and diarrhea",
    "loose and pasty stool",
    "melena",
    "nausea and vomiting",
    "hematemesis",
    "loose stool",
    "abdominal pain and diarrhea",
    "dysphagia",
    "abdominal rebound tenderness",
    "abdominal pain and distension",
    "abdominal wound",
    "soft abdomen with tenderness",
]

In [64]:
symptoms_df["gastro_int_symptom"] = symptoms_df["SYMPTOM_NAME"].apply(
    lambda symptom_name: 1 if symptom_name in gastrointestinal_symptoms else 0
)

In [65]:
symptoms_df.head()

SUBJECT_ID  HADM_ID          SYMPTOM_NAME SYMPTOM_ATTRIBUTE  \
0       12104   112523                 cough                 +   
1       12104   112523         good appetite                 +   
2       12104   112523           good spirit                 +   
3       12104   112523              listless                 +   
4       12104   112523  regular heart rhythm                 +   

   gastro_int_symptom  
0                   0  
1                   1  
2                   0  
3                   0  
4                   0

In [66]:
# filter all rows with a present gastro-intestinal symptom so all HADM_IDs present here have at least one such symptom, as missing
# HADM_IDs do not
symptoms_df = symptoms_df[symptoms_df["gastro_int_symptom"] == 1]

In [67]:
symptoms_df

SUBJECT_ID  HADM_ID    SYMPTOM_NAME SYMPTOM_ATTRIBUTE  \
1            12104   112523   good appetite                 +   
5            12104   112523  abdominal soft                 +   
35           12167   112587  abdominal soft                 +   
54           12054   112478  abdominal soft                 +   
69           12104   112523  abdominal soft                 +   
...            ...      ...             ...               ...   
402077       13253   113827  abdominal soft                 +   
402119       13253   113827  abdominal soft                 +   
402130       13615   114170  abdominal soft                 +   
402132       13615   114170  abdominal soft                 +   
402135       13615   114170  abdominal soft                 +   

        gastro_int_symptom  
1                        1  
5                        1  
35                       1  
54                       1  
69                       1  
...                    ...  
402077                   1  
402119                   1  
402130                   1  
402132                   1  
402135                   1  

[36246 rows x 5 columns]

In [68]:
hadm_id_to_symptom_flg = {
    hadm_id: symptom_flg
    for hadm_id, symptom_flg in zip(
        symptoms_df["HADM_ID"], symptoms_df["gastro_int_symptom"]
    )
}
icu_stay_df_merged["showed_gastro_intestinal_symptoms"] = icu_stay_df_merged[
    "HADM_ID"
].map(hadm_id_to_symptom_flg)
# TODO: possible to combine it with the mapping step above?
icu_stay_df_merged["showed_gastro_intestinal_symptoms"] = icu_stay_df_merged[
    "showed_gastro_intestinal_symptoms"
].fillna(0)

In [70]:
icu_stay_df_merged.to_csv("./icustay_min.csv", index=False)

## Add symptoms
We are now ready to add symptoms: We're not going to add all symptoms (probably) and not grouping by the symptom complex/category, as this could introduce bias and decreases reproducibilty.

### Add most common symptoms and set flag whether it has been present during the ICU stay or not

In [71]:
symptoms_df_ = pd.read_csv("./V1.1.0/EMR_SYMPTOMS.csv")
symptoms_df_.drop(
    ["ROW_ID", "EMR_ID", "RECORDTIME", "SYMPTOM_NAME_CN"], axis=1, inplace=True
)

In [72]:
symptoms_df_ = symptoms_df_.loc[symptoms_df_.SYMPTOM_ATTRIBUTE == "+"]
# side note: there are 13449 different HADMs in the dataset but only 892 HADMs have some positive symptoms
sympt_grp_by_hadm_nunique = symptoms_df_.groupby("SYMPTOM_NAME")["HADM_ID"].nunique()
sympt_grp_by_hadm_unique = symptoms_df_.groupby("SYMPTOM_NAME")["HADM_ID"].unique()

In [73]:
# filter all symptoms that are in at least 50% of the HADMs with at least one symptom
most_common_symptoms = sympt_grp_by_hadm_nunique.loc[
    (sympt_grp_by_hadm_nunique / 892) >= 0.50
]

In [74]:
# get all HADMs per symptom as filtered above
most_common_symptoms = sympt_grp_by_hadm_unique.loc[
    sympt_grp_by_hadm_unique.index.isin(most_common_symptoms.index)
]

In [75]:
icu_stay_df_merged = pd.read_csv("./merged_icustay_no_symptoms.csv")

In [76]:
most_common_symptoms = most_common_symptoms.rename("showed_{}".format)
sympt_idx = most_common_symptoms.index.to_list()
l = []
for symptom in sympt_idx:
    d = np.where(
        icu_stay_df_merged["HADM_ID"].isin(most_common_symptoms[symptom]), 1, 0
    )
    l.append(d)

In [77]:
most_common_symptoms

SYMPTOM_NAME
showed_abdominal soft                    [112523, 112587, 112478, 112568, 112549, 11223...
showed_anhelation                        [112549, 112559, 112607, 112604, 112568, 11263...
showed_arrhythmia                        [112550, 112511, 112549, 112598, 112562, 11258...
showed_atelectasis                       [112550, 112511, 112549, 112523, 112626, 11258...
showed_body infection                    [112550, 112511, 112549, 112586, 112598, 11283...
showed_cardiac arrest                    [112559, 112598, 112629, 112628, 112673, 11265...
showed_cough                             [112523, 112587, 112568, 112559, 112431, 11254...
showed_erythra                           [112550, 112478, 112511, 112549, 112523, 11248...
showed_extremity necrosis                [112550, 112511, 112549, 112586, 112598, 11283...
showed_fever                             [112523, 112559, 112550, 112478, 112511, 11256...
showed_good spirit                       [112523, 112587, 112559, 112478, 112

In [78]:
# Map new columns as keys and new values as values
col_val_mapping = dict(zip(sympt_idx, l))
# Unpack new column/new value pairs and assign them to the data frame
icu_stay_df_merged = icu_stay_df_merged.assign(**col_val_mapping)

In [80]:
icu_stay_df_merged.drop("Unnamed: 0", inplace=True, axis=1)
icu_stay_df_merged.to_csv("./merged_icustay_all.csv", index=False)